apps/language_server/lib/language_server/providers/formatting.ex

```diff
@@ -53,21 +53,49 @@ defmodule ElixirLS.LanguageServer.Providers.Formatting do
 
   defp can_format?(_uri, _project_dir), do: false
 
-  def should_format?(file_uri, project_dir, inputs) when is_list(inputs) do
+  defp should_format?(file_uri, project_dir, inputs) when is_list(inputs) do
     file_path = file_uri |> SourceFile.abs_path_from_uri()
 
-    inputs
-    |> Stream.flat_map(fn glob ->
-      [
-        Path.join([project_dir, glob]),
-        Path.join([project_dir, "apps", "*", glob])
-      ]
-    end)
-    |> Stream.flat_map(&Path.wildcard(&1, match_dot: true))
-    |> Enum.any?(&(file_path == &1))
+    if File.exists?(file_path) do
+      formatter_dir = find_formatter_dir(project_dir, file_path)
+
+      inputs
+      |> Stream.flat_map(&Path.wildcard(Path.join([formatter_dir, &1]), match_dot: true))
+      |> Enum.any?(&(file_path == &1))
+    else
+      # If the file doesn't actually exist on disk, we can't currently check
+      # whether we should format it via Path.wildcard/2, as that operation is
+      # dependent on the filesystem. So just assume that we should.
+      true
+    end
   end
 
-  def should_format?(_file_uri, _project_dir, _inputs), do: true
+  defp should_format?(_file_uri, _project_dir, _inputs), do: true
+
+  # Finds the deepest directory that contains file_path, that also contains a
+  # .formatter.exs. It's possible, though unlikely, that the .formatter.exs we
+  # find is not actually linked to the project_dir via the :subdirectories
+  # option in the top-level .formatter.exs. Currently, that edge case is
+  # glossed over.
+  defp find_formatter_dir(project_dir, file_path) do
+    parts = Path.split(file_path)
+    project_parts = Path.split(project_dir)
+
+    unless List.starts_with?(parts, project_parts) do
```

Comment:
This may not work well with symlinks

Reply:
I'll have a think about this, although if there's not an obvious solution I'd say that merging this is still far better than the status quo.